packages/ra-ui-materialui/src/field/ArrayField.tsx

```diff
@@ -134,39 +131,37 @@ export class ArrayField extends Component<
                   }, {}),
                   ids: list.map(JSON.stringify),
               };
-    }
+    };
 
-    render() {
-        const {
-            addLabel,
-            basePath,
-            children,
-            record,
-            sortable,
-            source,
-            fieldKey,
-            ...rest
-        } = this.props;
-        const { ids, data } = this.state;
+    const { ids: initialIds, data: initialData } = record
+        ? getDataAndIds(record, source, fieldKey)
+        : initialState;
+    const [ids, setIds] = useState(initialIds);
+    const [data, setData] = useState(initialData);
 
-        // @ts-ignore
-        return cloneElement(Children.only(children), {
-            ids,
-            data,
-            loading: false,
-            basePath,
-            currentSort: {},
-            ...rest,
-        });
-    }
-}
-
-const EnhancedArrayField = pure<FieldProps>(ArrayField);
+    useEffect(() => {
+        const { ids, data } = getDataAndIds(record, source, fieldKey);
+        setIds(ids);
+        setData(data);
+    }, [record]);
 
+    // @ts-ignore
+    return cloneElement(Children.only(children), {
+        ids,
+        data,
+        loading: false,
+        basePath,
+        currentSort: {},
+        ...rest,
+    });
+};
+// @ts-ignore
+const EnhancedArrayField = memo<FieldProps>(ArrayField);
+// @ts-ignore
 EnhancedArrayField.defaultProps = {
     addLabel: true,
 };
-
+// @ts-ignore
```

Comment:
shouldn't be necessary

Reply:
The same.

Reply:
I rollback to use recompose.pure insead of memo. 2 ts-ignore gone.